### PR TITLE
Allow RPC calls using block hash or number of latest block

### PR DESCRIPTION
**Pull request by cptartur**
## Usage related changes

- Currently, in many places in the RPC, it is required that the users uses `"latest"` as block hash/number. With this PR it is also possible to use a `block_hash` or `block_number` if they belong to the latest block.
- This PR allows making calls  with block tag `"pending"` block, though it is still treated as a `"latest"` block (up to discussion if this should be included).
- Changed fixtures using  pytest.mark.parametrize usages with `indirect=True`, to parametrized fixtures.

## Development related changes

None

## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/lint.sh`
- [x] Performed code self-review
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Documented the changes
- [ ] Linked the issues which this PR resolves
- [x] Updated the tests
- [ ] All tests are passing

### CODE REVIEW

File: starknet_devnet/blueprints/rpc/utils.py [hunk 1]
```diff
@@ -47,15 +51,30 @@ def get_block_by_block_id(block_id: BlockId) -> dict:
...
-    Assert block_id is "latest" and throw RpcError otherwise
+    Assert block_id is "latest"/"pending" or a block hash or number of "latest"/"pending" block and throw RpcError otherwise
     """
-    if block_id != "latest":
-        raise RpcError(
-            code=-1,
-            message="Calls with block_id != 'latest' are not supported currently.",
-        )
+    last_block = state.starknet_wrapper.blocks.get_last_block()
+    print(type(block_id), block_id)
```
Comment (FabijanC) [inline starknet_devnet/blueprints/rpc/utils.py:59]: print?

File: starknet_devnet/blueprints/rpc/utils.py [hunk 2]
```diff
@@ -47,15 +51,30 @@ def get_block_by_block_id(block_id: BlockId) -> dict:
...
     """
-    Assert block_id is "latest" and throw RpcError otherwise
+    Assert block_id is "latest"/"pending" or a block hash or number of "latest"/"pending" block and throw RpcError otherwise
     """
-    if block_id != "latest":
-        raise RpcError(
-            code=-1,
-            message="Calls with block_id != 'latest' are not supported currently.",
-        )
+    last_block = state.starknet_wrapper.blocks.get_last_block()
```
Comment (FabijanC) [inline starknet_devnet/blueprints/rpc/utils.py:58]: This last block retrieval can be moved inside of the `if isinstance(block_id, dict)` clause

File: starknet_devnet/blueprints/rpc/utils.py [hunk 3]
```diff
@@ -47,15 +51,30 @@ def get_block_by_block_id(block_id: BlockId) -> dict:
...
+            if int(block_id["block_number"]) == last_block.block_number:
+                return
+
+    if isinstance(block_id, str):
+        if block_id in ("latest", "pending"):
+            return
+
+    raise RpcError(
+        code=-1,
+        message="Calls with block_id != 'latest' are not supported currently.",
```
Comment (FabijanC) [inline starknet_devnet/blueprints/rpc/utils.py:76]: The error message should also be modified, right? If I may suggest extracting the valid block_ids to a variable and using it for both the actual checking and the message construction.

**Review by FabijanC — CHANGES_REQUESTED** [review]
Thanks for the contribution!

1. Accepted block IDs can be documented in `page/docs/guide/json-rpc-api.md`.
2. Currently only `"latest"` is tested. Shouldn't providing `"pending"` also be tested?

File: starknet_devnet/blueprints/rpc/utils.py [hunk 4]
```diff
@@ -47,15 +51,30 @@ def get_block_by_block_id(block_id: BlockId) -> dict:
...
-    if block_id != "latest":
-        raise RpcError(
-            code=-1,
-            message="Calls with block_id != 'latest' are not supported currently.",
-        )
+    last_block = state.starknet_wrapper.blocks.get_last_block()
+    print(type(block_id), block_id)
+
+    if isinstance(block_id, dict):
+        if "block_hash" in block_id:
```
[the remark this answers is not in the document]
Comment (cptartur) [reply starknet_devnet/blueprints/rpc/utils.py:62]: I think it's not handled anywhere, I'll include a test for this and a proper check.

**Comment by cptartur** [comment]
1. Will be added
2. I wasn't sure if we want to include "pending", for me it seems like a logical addition since `"pending"` is supported in gateway as far as I know. I'll extend the test to cover pending block.

**Comment by cptartur** [comment]
Addressed the comments. I've also unchecked "tests are passing" because `test/test_postman.py::test_postman` suddenly stopped working. I suspect it's a problem with some local configuration on my end so if tests could be run on your end we could verify that.

File: page/docs/guide/json-rpc-api.md [hunk 5]
```diff
@@ -31,3 +31,31 @@ Response:
 Methods currently not supported:
 
 - `starknet_protocolVersion` - will be removed in a future version of the specification
+
+Methods that require a `block_id` only support ids of the `latest` or `pending` block:
+
+```json
```
Comment (FabijanC) [inline page/docs/guide/json-rpc-api.md:37]: `json` doesn't support comments, so if you could change the language to `js` or `ts`.

File: page/docs/guide/json-rpc-api.md [hunk 6]
```diff
@@ -31,3 +31,31 @@ Response:
 Methods currently not supported:
 
 - `starknet_protocolVersion` - will be removed in a future version of the specification
+
+Methods that require a `block_id` only support ids of the `latest` or `pending` block:
```
Comment (FabijanC) [inline page/docs/guide/json-rpc-api.md:35]: Can you point out that these tags refer to the same block?

File: test/rpc/test_rpc_call.py [hunk 7]
```diff
@@ -52,6 +52,29 @@ def test_call_raises_on_incorrect_contract_address():
     assert ex["error"] == {"code": 20, "message": "Contract not found"}
 
 
+@pytest.mark.usefixtures("run_devnet_in_background", "deploy_info")
+def test_call_raises_on_both_hash_and_number():
+    """
+    Call contract with incorrect address
```
Comment (FabijanC) [inline test/rpc/test_rpc_call.py:58]: Docstring not corresponding to test.

**Review by FabijanC — CHANGES_REQUESTED** [review]
Cloned your fork, postman test is passing.

**Review by FabijanC — APPROVED** [review]

**Pull request merged**